libraries/AP_UAVCAN/AP_UAVCAN.cpp

```diff
@@ -319,6 +322,80 @@ void AP_UAVCAN::init(uint8_t driver_index, bool enable_filters)
     debug_uavcan(2, "UAVCAN: init done\n\r");
 }
 
+// send ESC telemetry messages over MAVLink
+void AP_UAVCAN::send_esc_telemetry_mavlink(uint8_t mav_chan)
+{
+
+    WITH_SEMAPHORE(_esc_sem);
```

Comment:
I like the way that KDECAN makes a copy of the data and works on that - vastly narrows the scope of holding that semaphore.  We really don't want to hold up the CAN thread while we stuff around assembling mavlink packets.  However, 200 bytes on the stack is also rather unpalatable (18 servos @14 bytes/esc entry).

Perhaps we could hold the semaphore while populating the arrays?  (and checking the groups)

i.e. take it just before checking `_escs_data[esc_id].available` on line 363

Reply:
KDECAN works on a copy of the data, but it holds the semaphore while making the copy. The fact that it holds the semaphore across narrower scope does not necessarily mean it holds it for **less time** which is what matters.
However, I do agree that the semaphore in my case needs to be moved.